### PR TITLE
addrman_tests

### PR DESCRIPTION
**Pull request by ghost**
**src/test/addrman_tests.cpp**
//after changing ports back from 12024 to 8333 addrman_tests pass successful.

On the screenshot you can see it pass addrman_tests.cpp
Screenshot from 2024-02-24 12-04-26

### CODE REVIEW

**Comment by ghost** [comment]
v7.17.3
https://github.com/DigiByte-Core/digibyte/blob/258afaceba07a16b8cf3acefc5bf2ce70d213b68/src/policy/feerate.cpp#L13

**Comment by JaredTate** [comment]
Thanks for doing this! A couple of quick questions, other than the fact it was using `size_t` in 7.17.3 was there any other reason to change this? I think for cross-platform consistency we should use `uint32_t`. 

Chat GPT summary is helpful to understand why:

> In C++, uint32_t and size_t are both fundamental types used for different purposes:
> 
> uint32_t: This is an unsigned 32-bit integer type defined in the <cstdint> header. It's explicitly sized to be 32 bits wide, regardless of the platform. It is useful when you need a specific width integer, for example, when dealing with data formats or protocols that require 32-bit unsigned integers.
> 
> size_t: This is an unsigned integer type defined in the <cstddef> header. It's typically used to represent sizes and indices of objects in memory. Its size may vary depending on the platform; however, it's guaranteed to be large enough to represent the size of the largest object supported by the implementation. This means it's usually the most appropriate type to use when dealing with sizes of arrays, memory allocations, or indices within arrays.
> 
> In summary:
> 
> uint32_t is explicitly 32 bits wide and is used when you need a fixed-width unsigned integer.
> size_t is platform-dependent and is typically used for representing sizes and indices within memory.
> 

Please let me know if I am missing something else with this, but I think BTC devs made the change for cross platform consistency.

**Comment by ghost** [comment]
> Thanks for doing this! A couple of quick questions, other than the fact it was using `size_t` in 7.17.3 was there any other reason to change this? I think for cross-platform consistency we should use `uint32_t`.

I saw a problem in test/amount_tests.cpp. The size of DigiByte MAX_MONEY was 21e17. Bitcoin MAX_MONEY is 21e15.
It only worked when i did. 
```
size_t MAX_MONEY2;
MAX_MONEY2 = MAX_MONEY / 1000;
CFeeRate(MAX_MONEY2, std::numeric_limits<uint32_t>::max()).GetFeePerK(); 
```
https://github.com/DigiByte-Core/digibyte/blob/dd95b35baebe2337c6638c6ce9b8069b24fedbda/src/test/amount_tests.cpp#L88

Then i saw in policy/feerate.h and policy/feerate.cpp uint32_t. And changed it just to be sure. saw the same at Litecoin and DigiByte 7. 

Here num_bytes is converted immediately to int64_t
https://github.com/DigiByte-Core/digibyte/blob/dd95b35baebe2337c6638c6ce9b8069b24fedbda/src/policy/feerate.cpp#L12-L14

But of course i have changed it back to uint32_t. uint32_t is better. Thank you!!

**Comment by ghost** [comment]
in **src/test/amount_tests.cpp** i also change **int64_t** to **uint32_t**.

File: src/policy/feerate.cpp
Comment (gto90) [inline src/policy/feerate.cpp]: Minor nitpick here @Jongjan88 , but could you fix the formatting here and line 19?

File: src/test/amount_tests.cpp
Comment (gto90) [inline src/test/amount_tests.cpp]: @Jongjan88 and @JaredTate, I am concerned about this change.  Normally, I would not expect changes to our test code but rather the underlying code being tested for the respective tests to pass.

The proposed change, however, alters the test to use a boolean expression result (either 0 or 1) as the first argument to CFeeRate. This means the test no longer checks the behavior of CFeeRate with large numerical values but rather with the smallest possible non-negative values (0 or 1).

This shift in testing could potentially miss issues or behaviors that only manifest with larger values.  Changing a test is unusual unless the underlying code or the specifications have changed.

@Jongjan88, can you provide some more detail on this change?

**Review by gto90 — CHANGES_REQUESTED** [review]
I left some comments for @JaredTate and @Jongjan88 to review.

File: src/test/amount_tests.cpp
Comment (ghost) [reply src/test/amount_tests.cpp]: I think you are right. We should not change the test if possible.

When i do this. The code is working.
```
size_t = MAX_MONEY2
MAX_MONEY2 = MAX_MONEY / 1000;
CFeeRate(MAX_MONEY2, std::numeric_limits<uint32_t>::max()).GetFeePerK();
```

**Update**
I remove this line CFeeRate(MAX_MONEY <= uint32_t(std::numeric_limits<uint32_t>::max())).GetFeePerK();

File: src/policy/feerate.cpp
Comment (gto90) [inline src/policy/feerate.cpp]: @JaredTate and @Jongjan88, will performing an assertion in production code cause any performance issues?  This strengthens the integrity of these codes, as does the code on line 13, but I would like to know if these assertions should move to a unit test.

**Comment by ycagel** [comment]
@Jongjan88 Are these all the correct changes and should other devs be reviewing this?

**Comment by ghost** [comment]
> @Jongjan88 Are these all the correct changes and should other devs be reviewing this?

Yes.

**Review by gto90 — DISMISSED** [review]
cACK.

This looks good now @Jongjan88 , but I do have a question about placing the assertions in `feerate.cpp`.  While this does make this code more guarded, I wonder what the performance impact will be?  I will have to defer to @JaredTate .  Conceptually, this PR looks ok though.

**Review by ycagel — DISMISSED** [review]
cACK. I agree with @gto90 that we need to get a further review from @JaredTate.

**Comment by JaredTate** [comment]
Ok, I had reviewed and compiled it yesterday but now there are only 2 changes. Is that correct?  Others were reverted?
Screenshot 2024-02-27 at 3 16 05 PM

**Comment by gto90** [comment]
> Ok, I had reviewed and compiled it yesterday but now there are only 2 changes. Is that correct? Others were reverted? Screenshot 2024-02-27 at 3 16 05 PM

That is correct @JaredTate !  I believe they were reverted because they were changes to the test code which obfuscated underlying issues we are actually having in the code.

File: src/policy/feerate.cpp
Comment (ghost) [reply src/policy/feerate.cpp]: **I remove freerate.cpp from the pr.**

Here in the latest version of litecoin.
https://github.com/litecoin-project/litecoin/blob/8925697f15f716da94c510c63b2fe7dac852cb0b/src/policy/feerate.cpp#L32-L34

**ChatGPT**
For example, if the uint32_t value is 4,294,967,2964,294,967,296 (which is 232232), and you attempt to convert it to an int64_t, which cannot represent that value, you will encounter problems such as overflow, where the resulting int64_t will wrap around to a negative value.

Don't know if it gives a serious performance hit. But it is more secure. Maybe something to look at in the future.
We can look at Bitcoin. But we have bigger fees and coin size than bitcoin.

-------------------

(https://github.com/DigiByte-Core/digibyte/issues/164)
Most likely this won't solve it. But there are problems that still pass the test. That's why I think an extra line of code can sometimes help.

Comment (gto90) [reply src/policy/feerate.cpp]: I agree with you @Jongjan88 , I just know that those assertions do add a bit of overhead but not sure how much of an impact it is.  I don't dislike this change; just wanted to see what @JaredTate thought about it.

I think it would be ok for you to open a separate PR for that however.

**Review by gto90 — COMMENTED** [review]
@Jongjan88 I see that we use 12024 still in a couple of files, should they also be changed?

- nodes_main.txt
- windows-setup.md
- chainparams.cpp
- netbase.h
- net.cpp
- net_tests.cpp
- netbase_tests.cpp

**Comment by gto90** [comment]
> @Jongjan88 I see that we use 12024 still in a couple of files, should they also be changed?
> 
> * nodes_main.txt
> * windows-setup.md
> * chainparams.cpp
> * netbase.h
> * net.cpp
> * net_tests.cpp
> * netbase_tests.cpp

@Jongjan88 , we do not necessarily have to change them in this PR to be clear.  But we should make sure that our ports are consistent across the code base.

**Comment by ghost** [comment]
> > @Jongjan88 I see that we use 12024 still in a couple of files, should they also be changed?
> > 
> > * nodes_main.txt
> > * windows-setup.md
> > * chainparams.cpp
> > * netbase.h
> > * net.cpp
> > * net_tests.cpp
> > * netbase_tests.cpp
> 
> @Jongjan88 , we do not necessarily have to change them in this PR to be clear. But we should make sure that our ports are consistent across the code base.

@gto90 

I saw them. I think some need to be changed. chainparams.cpp not ofc.
But haven't come across a test yet that failed due to the port. except addrman_tests
In the future it would be nice that the test also use 12024. But for now i don't think it is a problem to let them on port 8333.
Like before PR https://github.com/DigiByte-Core/digibyte/pull/166

https://github.com/DigiByte-Core/digibyte/issues/178#issue-2153039381
Need some help on test 1 amount_tests.cpp(88) @JaredTate 

@gto90 
Haha. My next PRs are definitely going to be better. last 2 PRs were a disaster! Sorry for this.

**Review by gto90 — APPROVED** [review]
utACK

Thank you @Jongjan88

**Comment by JaredTate** [comment]
No worries on this, so now it's just port changes correct? I am trying to free up a couple of full days to really dive in on this. I have had other commitments to finish over this past week. I should free up the next few days.

**Comment by ghost** [comment]
> No worries on this, so now it's just port changes correct? I am trying to free up a couple of full days to really dive in on this. I have had other commitments to finish over this past week. I should free up the next few days.

Nice. Yes after this port change in addrman_tests. addrman_tests will not give 3 errors. And pass successful.

**Review by JaredTate — COMMENTED** [review]
ACK. Make sense, thanks for doing this. It compiles & runs good. On to the next!

**Review by ycagel — APPROVED** [review]
cACK. Thanks for your efforts folks!

**Review by JaredTate — APPROVED** [review]

**Pull request merged**
